### DISCUSSION TITLE
Manually trigger menu then automatically select and insert first result

### DISCUSSION THREAD

**Discussion by gbishop** [hrsh7th/nvim-cmp#1424] [Q&A]
I'd like to  manually trigger the completion menu and have the first result both selected and inserted. I can get it selected but I can't get that first result automatically inserted. 

I've mapped the TAB key to trigger the menu. In my dreams I'd be typing along and hit TAB to have the correct result appear with no other keys required. My current configuration requires that I hit TAB twice; once to open the menu and once to select and insert the result. 

EDIT: I don't simply want to insert the first choice and be done with it; I want to reserve the possibility of hitting TAB again to select a later result. It would work just like the two TAB sequence but with only 1 TAB. 

How can I open the menu and both select and insert the first item with a single key?

I included the minimal configuration below. Thanks for any suggestions.

```
if has('vim_starting')
  set encoding=utf-8
endif
scriptencoding utf-8

if &compatible
  set nocompatible
endif

let s:plug_dir = expand('/tmp/plugged/vim-plug')
if !filereadable(s:plug_dir .. '/plug.vim')
  execute printf('!curl -fLo %s/autoload/plug.vim --create-dirs https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim', s:plug_dir)
end

execute 'set runtimepath+=' . s:plug_dir
call plug#begin(s:plug_dir)
Plug 'hrsh7th/nvim-cmp'
Plug 'hrsh7th/cmp-buffer'
Plug 'hrsh7th/cmp-nvim-lsp'
Plug 'hrsh7th/vim-vsnip'
Plug 'neovim/nvim-lspconfig'
call plug#end()
PlugInstall | quit

" Setup global configuration. More on configuration below.
lua << EOF

local has_words_before = function()
	unpack = unpack or table.unpack
	local line, col = unpack(vim.api.nvim_win_get_cursor(0))
	return col ~= 0
		and vim.api
				.nvim_buf_get_lines(0, line - 1, line, true)[1]
				:sub(col, col)
				:match("%s")
			== nil
end

local cmp = require "cmp"
cmp.setup {
	completion = {
		autocomplete = false,
		completeopt = "menu,menuone,preview",
	},

  snippet = {
    expand = function(args)
      vim.fn["vsnip#anonymous"](args.body)
    end,
  },

  mapping = {
    ["<Tab>"] = cmp.mapping(function(fallback)
			if cmp.visible() then
				cmp.select_next_item()
			elseif has_words_before() then
				cmp.complete()
			else
				fallback()
			end
		end, { "i", "s" }),
  },

  sources = cmp.config.sources({
    { name = "buffer" },
  }),
}
EOF
```

**Comment by fitrh** [comment]
I think you can use the event, see `:h cmp.event:on`

Something like this

```lua
cmp.event:on('menu_opened', function()
        cmp.confirm({ select = true })
end)
```

**Reply by gbishop** [reply]
See the EDIT above. I tried that with cmp.select_next_item and it restarted the menu as if I had typed in the first result. 

I don't want to confirm, I want to still be able to choose other results.

And thanks for the answer. I'm sorry about leaving out that detail. It came to me as I was heading to bed that I left that out.

**Reply by fitrh** [reply]
You can try setting the select behavior instead of `confirm`

```lua
cmp.select_next_item({ behavior = cmp.SelectBehavior.Insert })
```

**Reply by gbishop** [reply]
I've got it working using feedkeys in the event:on open callback. I'll post the solution later today.

**Comment by gbishop** [comment]
The solution below works just as I wanted. I can hit Tab to trigger a completion with the first result **selected** and **inserted**. If I want that result I can just keep typing the next symbol and the menu closes. If I don't want that item I can continue to hit Tab to move down the list. I'll likely include a cancel key and such but this alone works for most cases that interest me.

One key to complete in most cases and no menus popping up unbidden. This is the flow for me!

Thanks for the help.
gb

```lua
local has_words_before = function()
	unpack = unpack or table.unpack
	local line, col = unpack(vim.api.nvim_win_get_cursor(0))
	return col ~= 0
		and vim.api
				.nvim_buf_get_lines(0, line - 1, line, true)[1]
				:sub(col, col)
				:match("%s")
			== nil
end

local cmp = require "cmp"
cmp.setup {
	completion = {
		autocomplete = false,
		completeopt = "menu,menuone,noselect",
	},

  snippet = {
    expand = function(args)
      vim.fn["vsnip#anonymous"](args.body)
    end,
  },

  mapping = {
		["<Tab>"] = cmp.mapping(function(fallback)
			if cmp.visible() then
				cmp.select_next_item()
			elseif has_words_before() then
				cmp.complete()
			else
				fallback()
			end
		end, { "i", "s" }),
  },

  sources = cmp.config.sources({
    { name = "buffer" },
  }),
}

cmp.event:on("menu_opened", function()
    -- insert a Tab key to select and insert the first result
    vim.api.nvim_feedkeys('\t', 'i', false)
end)
```

**Reply by gbishop** [reply]
Nope. This method has problems with getting into loops. I'll work on it more tomorrow.

**Reply by azr4e1** [reply]
Hi!

I'm interested in this use case, I have your exact same workflow, did you manage to figure out the issue with getting into loops?

Thanks!

**Reply by gbishop** [reply]
I couldn't get it to work so I went to [mini.completion](https://github.com/echasnovski/mini.completion) with this configuration:
```lua
require("mini.completion").setup({
	delay = {
		completion = 10 ^ 7,
		info = 100,
		signature = 100,
	},
	lsp_completion = {
		source_func = "omnifunc",
		auto_setup = false,
	},
	set_vim_settings = false,
})

local keys = {
	["cr"] = vim.api.nvim_replace_termcodes("<CR>", true, true, true),
	["ctrl-y"] = vim.api.nvim_replace_termcodes("<C-y>", true, true, true),
	["ctrl-y_cr"] = vim.api.nvim_replace_termcodes("<C-y><CR>", true, true, true),
}

_G.has_words_before = function()
	unpack = unpack or table.unpack
	local line, col = unpack(vim.api.nvim_win_get_cursor(0))
	return col ~= 0
		and vim.api
				.nvim_buf_get_lines(0, line - 1, line, true)[1]
				:sub(col, col)
				:match("%s")
			== nil
end

vim.api.nvim_set_keymap(
	"i",
	"<Tab>",
	[[pumvisible() ? "\<C-n>" : v:lua._G.has_words_before() ? "\<C-Space>" : "\<Tab>"]],
	{ noremap = false, expr = true }
)

vim.api.nvim_set_keymap(
	"i",
	"<S-Tab>",
	[[pumvisible() ? "\<C-p>" : "\<S-Tab>"]],
	{ noremap = true, expr = true }
)
```
I believe this option is required as well:
```
vim.o.completeopt = "menu,menuone"
```

**Comment by Thanatermesis** [comment]
Oh! I also need this feature, or similar, yesterday I opened an issue about that: https://github.com/hrsh7th/nvim-cmp/issues/1529

**Comment by Thanatermesis** [comment]
@gbishop I think I found the solution: https://github.com/hrsh7th/nvim-cmp/issues/1529#issuecomment-1532582272

Tell me if you have an issue setting it up so I can share you my full conf so you can compare

**Reply by gbishop** [reply]
I'd like to see the full config please. My method using mini.complete is working pretty well but I would be interested in trying this.

**Reply by gbishop** [reply]
I don't have time to work on this now. I gave it a quick try and it didn't do what I wanted but I likely missed something. I'm happy with my mini.completion setup for now.
